File: leetcode_javascript/113.PathSum_II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function(root, targetSum) {
    const ans = [];
    traversal(root, [], targetSum);
    return ans;

    function traversal(root, path, targetSum) {
        if (root === null) return;
        path.push(root.val);

        if (root.left === null && root.right === null) {
            if (targetSum === root.val) ans.push(path);
            return;
        }

        targetSum -= root.val;
        if (root.left !== null) traversal(root.left, [... path], targetSum);
        if (root.right !== null) traversal(root.right, [... path], targetSum);
    }
};

File: leetcode_python/113. Path Sum II.py
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        def getPath(node, path, sum):  # path must be passed by value.
            if not node:
                return
            
            sum -= node.val
            path.append(node.val)
            if node.left is None and node.right is None:
                if sum == 0:
                    result.append(path)
                return
            
            getPath(node.left, path[:], sum)
            getPath(node.right, path[:], sum)
        
        result = []
        path = []
        getPath(root, path, targetSum)
        return result

/**
113. Path Sum II
Solved
Medium
Topics
conpanies iconCompanies

Given the root of a binary tree and an integer targetSum, return all root-to-leaf paths 
where the sum of the node values in the path equals targetSum. Each path should be returned 
as a list of the node values, not node references.

A root-to-leaf path is a path starting from the root and ending at any leaf node. 
A leaf is a node with no children.
 */
